### PR TITLE
Rewrite projectsView.js to fix pagination

### LINKED ISSUE #1759: Dashboard not paging correctly
## Expected Behaviour

When you scroll down the page new courses are bought in.

## Actual Behaviour

Only the initial 32 courses loaded are shown.

## Versions
  - Authoring Tool Version: 0.4
  - Framework Version: 2.2.2

### PR DESCRIPTION
**Pull request by taylortom**
Fixes #1759

There was a lot of redundant/superfluous code in here (I couldn't get my head around how the thing worked), so have cleaned up a bit to help with readability.

### CODE REVIEW

**Comment by dancgray** [comment]
Tried with over 50 courses and showed them all however resizing the dashboard window only renders one course.

File: frontend/src/modules/projects/views/projectsView.js [hunk 1]
```diff
@@ -1,280 +1,182 @@
...
-      return { title: pattern};
+    doLayout: function(layout) {
+      var layouts = ["grid", "list"];
+      if(_.indexOf(layouts, layout) === -1) {
+        return;
+      }
+      var classSuffix = '-layout';
+      this.getProjectsContainer()
+        .removeClass(layouts.join(classSuffix + ' ') + classSuffix)
+        .addClass(layout + '-layout');
```
Comment (tomgreenfield) [inline frontend/src/modules/projects/views/projectsView.js]: Not crucial to this PR but would you be privvy to using a data attribute for this instead? Could be cleaner / more flexible.

For example:
```js
this.$el.attr('data-layout', layout);
```
Earlier on in the file we would set the default:
```js
attributes: {
  'data-layout': 'grid'
}
```

[`projects.less`](https://github.com/adaptlearning/adapt_authoring/blob/issue/1759/frontend/src/modules/projects/less/projects.less#L41-L114) and [`projects.hbs`](https://github.com/adaptlearning/adapt_authoring/blob/issue/1759/frontend/src/modules/projects/templates/projects.hbs#L5) would need tweaks too.

File: frontend/src/modules/projects/views/projectsView.js [hunk 2]
```diff
@@ -1,280 +1,182 @@
...
-    filterBySearchInput: function (filterText) {
-      this.filterText = filterText;
-      this.search = this.convertFilterTextToPattern(filterText);
-      this.setUserPreference('search', filterText);
-      this.updateCollection(true);
-    },
+    doFilter: function(text, tags, fetch) {
+      text = text || '';
+      this.filterText = text;
+      this.search = this.convertFilterTextToPattern(text || '');
```
Comment (tomgreenfield) [inline frontend/src/modules/projects/views/projectsView.js]: Is the OR operator actually needed here considering line 170?

File: frontend/src/modules/projects/views/projectsView.js [hunk 3]
```diff
@@ -1,280 +1,182 @@
...
-      // Set empty filters
-      this.filters = [];
-      this.tags = [];
-
-      this.collectionLength = 0;
-      this.shouldStopFetches = false;
+    initUserPreferences: function() {
+      var prefs = this.getUserPreferences();
 
+      this.doLayout(prefs.layout, false);
```
Comment (tomgreenfield) [inline frontend/src/modules/projects/views/projectsView.js]: `doLayout()` doesn't need the second parameter?

File: frontend/src/modules/projects/views/projectsView.js [hunk 4]
```diff
@@ -1,280 +1,182 @@
...
-          : null ;
-      this.updateCollection(true);
-    },
-
-    convertFilterTextToPattern: function(filterText) {
-      var pattern = '.*' + filterText.toLowerCase() + '.*';
-      return { title: pattern};
+    doLayout: function(layout) {
+      var layouts = ["grid", "list"];
+      if(_.indexOf(layouts, layout) === -1) {
```
Comment (tomgreenfield) [inline frontend/src/modules/projects/views/projectsView.js]: Use native `indexOf()`? Underscore version only needed for IE8 support.

`_.contains()` would be the only real beneficial one if you wanted readability.

**Comment by taylortom** [comment]
Thanks for the review @tomgreenfield, have actioned your changes and fixed @dancgray's issue.

File: frontend/src/modules/projects/views/projectsView.js [hunk 5]
```diff
@@ -1,280 +1,198 @@
...
-        if ((scrollableHeight-containerHeight) - scrollTop < 30) {
-          if (!this.isCollectionFetching) {
-            this.isCollectionFetching = true;
-            this.lazyRenderCollection();
-          }
-        }
-      }, this));
+    onResize: function() {
+      // we don't want to re-initialise for _every_ resize event
+      if(this.resizeTimer) {
```
Comment (lc-thomasberger) [inline frontend/src/modules/projects/views/projectsView.js]: could you use _.throttle for this?

File: frontend/src/modules/projects/views/projectsView.js [hunk 6]
```diff
@@ -1,280 +1,198 @@
...
-      var locationTitleHeight = $('.location-title').outerHeight();
-      var windowHeight = $(window).height();
-      var actualHeight = windowHeight - (navigationHeight + locationTitleHeight);
-      this.$el.css('height', actualHeight);
-    },
+      this.listenTo(this.collection, 'add', this.appendProjectItem);
 
-    checkIfCollectionIsEmpty: function() {
-      this.$('.no-projects').toggleClass('display-none', this.collection.length > 0);
+      $('.contentPane').scroll(_.bind(this.doLazyScroll, this));
```
Comment (lc-thomasberger) [inline frontend/src/modules/projects/views/projectsView.js]: You may also use a throttled function for the scroll callback.

File: frontend/src/modules/projects/views/projectsView.js (hunk 5 above)
Comment (taylortom) [reply frontend/src/modules/projects/views/projectsView.js]: Yep, certainly could. Will have a look :)

**Comment by taylortom** [comment]
@dancgray @tomgreenfield are you happy for this to go in with the latest changes?

**Pull request merged**